New NVENC config page

ns6089 opened the pull request:
## Description
<!--- Please include a summary of the changes. --->

This updates the page to remove confusing, irrelevant and harmful options and adds new relevant ones,

Also changes the default options slightly:
- P1 is now the default (previously P4). P4 is significantly slower, while providing no real benefits for local streaming.
- Faster two-pass variant is enabled by default, without it vbr/hrd constraints are periodically overshoot by up to 2x. Also improves quality.
- HAGS scheduling priority freeze workaround is now disabled by default since it can significantly affect capture performance.

### Screenshot
<!--- Include screenshots if the changes are UI-related. --->

![2023-09-03 13_36_24-Sunshine — Mozilla Firefox](https://github.com/LizardByte/Sunshine/assets/61738816/388c81fb-0fc8-4a51-b0e6-99848c225110)

### Issues Fixed or Closed
<!--- Close issue example: `- Closes #1` --->
<!--- Fix bug issue example: `- Fixes #2` --->
<!--- Resolve issue example: `- Resolves #3` --->


## Type of Change
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Dependency update (updates to dependencies)
- [ ] Documentation update (changes to documentation)
- [ ] Repository update (changes to repository files, e.g. `.github/...`)

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated the in code docstring/documentation-blocks for new or existing methods/components

## Branch Updates
LizardByte requires that branches be up-to-date before merging. This means that after any PR is merged, this branch
must be updated before it can be merged. You must also
[Allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I want maintainers to keep my branch updated

ns6089 commented:
Should be ready to merge besides updating docs, but the page is now self-documented so maybe we can go forward with it without them. Will update docs slightly later when I have more time.

ReenigneArcher commented:
The docs are helpful when people ask how to do something, it's easier to give them a link to the option, than to try to guide them through the UI.

ns6089 commented:
> The docs are helpful when people ask how to do something, it's easier to give them a link to the option, than to try to guide them through the UI.

Like I said, I will update them later when I have time. If you want to block the PR until then, it's fine with me.

cgutman commented on src_assets/common/assets/web/config.html:
```diff
@@ -872,41 +872,73 @@ <h1 class="my-4">Configuration</h1>
+                  <a href="https://devblogs.microsoft.com/directx/hardware-accelerated-gpu-scheduling/">HAGS</a>
+                  is enabled, realtime priority is used and VRAM utilization is close to maximum.<br>
+                  Disabling this option lowers the priority to high sidestepping the freeze, but capture performance may suffer
+                  when GPU is heavily loaded.
```
The second sentence there is a bit hard to understand. Maybe something like: "Disabling this option lowers the priority to high, sidestepping the freeze at the cost of reduced capture performance when the GPU is heavily loaded."

ns6089 replied:
Copy-pasted, thank you.

ReenigneArcher reviewed (changes requested):
Mostly just some minor suggestions for the html, and a few questions on the C++.

ns6089 commented on src/nvenc/nvenc_config.h:
```diff
@@ -15,10 +15,10 @@ namespace nvenc {
   struct nvenc_config {
     // Quality preset from 1 to 7, higher is slower
-    unsigned quality_preset = 1;
+    int quality_preset = 1;
```
It's inconsequential here, but the config reader can't handle unsigned. And if I start fixing that reader, it's probably going to end up being another 1k lines refactoring.

ns6089 commented on src/video.cpp:
```diff
@@ -601,7 +601,7 @@ namespace video {
     PARALLEL_ENCODING | REF_FRAMES_INVALIDATION  // flags
   };
-#else
+#elif !defined(__APPLE__)
```
Seems like it. We certainly don't enable it, otherwise commenting out the declaration would break the compilation.

ns6089 commented on src_assets/common/assets/web/config.html:
```diff
@@ -872,41 +872,73 @@ <h1 class="my-4">Configuration</h1>
-          <option value="lossless">lossless -- lossless</option>
+        <label for="nvenc_preset" class="form-label">Quality preset</label>
+        <select id="nvenc_preset" class="form-select" v-model="config.nvenc_preset">
+          <option value="1">P1 (fastest, default)</option>
+          <option value="2">P2</option>
+          <option value="3">P3</option>
+          <option value="4">P4</option>
+          <option value="5">P5</option>
+          <option value="6">P6</option>
+          <option value="7">P7 (slowest)</option>
```
It doesn't make anything hurt though, probably no effect on 3D performance too. Only larger latency and lower peak throughput.

ReenigneArcher replied:
Okay, just trying to think of additional wording that would make people want to avoid this option.

ns6089 commented on src_assets/common/assets/web/config.html:
```diff
@@ -1061,10 +1093,10 @@ <h1 class="my-4">Configuration</h1>
+    "nvenc_preset": "1",
+    "nvenc_twopass": "quarter_res",
+    "nvenc_realtime_hags": "enabled",
+    "nvenc_h264_cavlc": "disabled",
```
Reordered.

ns6089 commented on src_assets/common/assets/web/config.html:
```diff
@@ -872,41 +872,73 @@ <h1 class="my-4">Configuration</h1>
+                </select>
+                <div class="form-text">Simpler form of entropy coding.<br>
+                  Needs around 10% more bitrate for same quality.<br>
+                  Only relevant for really old decoding devices.
```
Changed.

ns6089 commented on src_assets/common/assets/web/config.html:
```diff
@@ -872,41 +872,73 @@ <h1 class="my-4">Configuration</h1>
+          <option value="7">P7 (slowest)</option>
         </select>
+        <div class="form-text">Higher numbers improve compression (quality at given bitrate) at the cost of increased
+          encoding latency.<br>
```
Added the flavor.

ns6089 commented:
Rebased and updated docs. Should be ready to merge.

---
![2023-09-07 14_27_01-Sunshine — Mozilla Firefox](https://github.com/LizardByte/Sunshine/assets/61738816/ec3036c9-fd9b-4deb-b001-d873bbfe8c83)

---
![nvenc_docs](https://github.com/LizardByte/Sunshine/assets/61738816/adb458e1-22ca-4773-886f-2b1a8d920467)

ns6089 commented:
Affects https://github.com/LizardByte/Sunshine/issues/1031
